mitmproxy/net/http/request.py

```diff
@@ -367,13 +367,64 @@ def _get_urlencoded_form(self):
                 pass
         return ()
 
+    def _sanitized_form_data_replacement(self, form_data):
+        # check if unmodified content changes with URL encoding
+        url_decoded_content = urllib.parse.parse_qsl(self.content, keep_blank_values=True)
+        url_reencoded_content = urllib.parse.urlencode(url_decoded_content, doseq=True).encode()
+        if self.content == url_reencoded_content:
```

Comment:
There are 3 cases considered:

1. POST data isn't urlencoded at all 

Check: 
`
if self.content == urllib.parse.unquote(url_reencoded_content.decode()).encode():
`

triggering example:
`
self.content
b'date[year]=2017'
`

Tested with: 
`curl -k --proxy localhost:8080 'https://www.example.com/' -H 'Content-Type: application/x-www-form-urlencoded' --data 'date[year]=2017'`

2. POST data isn't urlencoded at all and contains empty key without equal sign

Check: 
`
elif self.content == urllib.parse.unquote(url_reencoded_content.decode()).replace("=&", "&").encode():
`

triggering example:
`
self.content
b'key_without_value&date[year]=2017'
`


Tested with: 
`curl -k --proxy localhost:8080 'https://www.example.com/' -H 'Content-Type: application/x-www-form-urlencoded' --data 'key_without_value&date[year]=2017'`

3. urldecoded POST data contains at least one empty key without equal sign: 

Check:
`
elif any("=" not in param for param in urllib.parse.unquote(self.content.decode()).split("&")):
`

triggers everytime a message is posted on facebook (https://www.facebook.com/ajax/updatestatus.php)

Tested by posting on facebook.


So all triggers work as intended on my side, but i agree that case 1 and case 2 are rare examples, which should be dropped in order to focus on the initial facebook issue.